managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2151,7 +2151,13 @@ public void operationFailed(ManagedLedgerException exception) {
 
                 mdEntry.triggerFailed(exception);
             }
-        });
+        };
+
+        if (State.NoLedger.equals(STATE_UPDATER.get(this))) {
+            persistPositionMetaStore(mdEntry, cb);
```

Comment:
If the BookKeeper cluster runs into read-only mode and millions of consumers keep acknowledging the mark-delete position to brokers, we will persist the mark-delete position to the meta store without any throttle policy, will it bring high pressure or risk for the meta store? especially when we use Zookeeper as the Pulsar cluster's meta store service.

Reply:
> If the BookKeeper cluster runs into read-only mode and millions of consumers keep acknowledging the mark-delete position to brokers, we will persist the mark-delete position to the meta store without any throttle policy, will it bring high pressure or risk for the meta store? especially when we use Zookeeper as the Pulsar cluster's meta store service.

Some users may want to set BK to read-only, and stop sending messages but consume messages.

However, pulsar does not support this yet, because even if only the message is consumed, there is still some BK write operations, here are some examples:
- If a transaction is used when consuming a message, a transaction log is also written
- Compaction tasks are still written to BK
- Topic policies depend on BK write
- https://github.com/apache/pulsar/issues/16691
- https://github.com/apache/pulsar/issues/16763
- https://github.com/apache/pulsar/issues/16042 

If we want to support the above scenario "Make BK read-only and only consume messages", we should need a large design.

Reply:
>  we will persist the mark-delete position to the meta store without any throttle policy

@hangc0276 Actually, we have a throttling policy that can be configured by `managedLedgerDefaultMarkDeleteRateLimit`. For improving this case, it looks like we'd better have a separate one for the zookeeper. So that users can persist to the cursor ledger for each second but persist to zookeeper for every 30 seconds.

Reply:
> @hangc0276 @poorbarcode Actually, we have a throttling policy that can be configured by managedLedgerDefaultMarkDeleteRateLimit. For improving this case, it looks like we'd better have a separate one for the zookeeper. So that users can persist to the cursor ledger for each second but persist to zookeeper for every 30 seconds.

Good idea. Since this solution should change `broker.conf` and `PersistencePolicies`, maybe a PIP is needed.

In the current PR, I've simpler this behavior to the following: persist to ZK only hits these two scenarios
- the cursor ledger is not null, persist to ZK after persisting to BK failed(the original behavior before this PR)
- the cursor ledger is null, and the `cursor.markDeletedPosition==managedLedger.LAC`. Since the ledger can not be created, no entry can be written now, so the variable `managedLedger.LAC` will not increase anymore. For per subscription, persist to ZK will only be executed once.

And in the future, we submit a PIP to add a rater limiter to further improve the behavior.

@codelipenghui @hangc0276   What do you think abort?

Reply:
Yes, it looks good to me. And users can also change `managedLedgerDefaultMarkDeleteRateLimit` to reduce the impact.

Reply:
Don' we need to consider `shouldPersistUnackRangesToLedger()` here, as well? If individualDeletedMessages is too big, bigger than maxUnackedRangesToPersistInMetadataStore, shouldn't we persist it to BK, instead of MetadataStore?